safe_upgrade/safe_upgrades/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+        .await
+        .map(|info| info.total_num_changes)
+        .map_err(add_stage(UpgradeStage::ObtainingInfo))?;
+
+    // 3) Install (upgrade) the new WASM. Loop until success or timeout. We can't retry directly
+    // here if we don't know what happened, since installation isn't idempotent. Instead, use the
+    // version number to determine if the upgrade went through.
+    loop {
+
+        let install_result = match wasm_module {
+            WasmModule::Bytes(ref wasm_bytes) => {
+                best_effort_install_single_chunk(target_id, wasm_bytes, &arg, should_retry).await
+            }
+            WasmModule::ChunkedModule(ref chunked) => {
+                best_effort_install_chunked(target_id, chunked, &arg, should_retry).await
+            }
+        };
+
+        match install_result {
+            Ok(()) => break,
+            // Note that for installation, unretriable errors include `SysUnknown`
+            // Try to figure out what happened using the version and retry if the version
+            // hasn't moved
+            Err(RetryError::StatusUnknown(ErrorCause::CallFailed(rejection)))
+                if !rejection.is_clean_reject() =>
+            {
+                let (new_version, mut recent_changes) = best_effort_canister_info(target_id, Some(1), should_retry)
+                    .await
+                    .map(|info| (info.total_num_changes, info.recent_changes))
+                    .map_err(add_stage(UpgradeStage::Installing))?;
+                if new_version <= version {
+                    ic_cdk::println!("Failed to upgrade {:?} and the version hasn't moved, retrying", target_id);
+                    continue;
+                } else {
+                    if new_version != version + 1 || recent_changes.len() != 1 {
+                        return Err(UpgradeError {
+                            stage: UpgradeStage::Installing,
+                            reason: UpgradeErrorReason::ConcurrentChangeDetected,
+                        });
+                    }
+                    if let Some(change) = recent_changes.pop() {
+                        if let ChangeDetails::CodeDeployment(dep) = change.details {
+                            let expected_hash: Vec<u8> = match wasm_module {
+                                WasmModule::Bytes(ref wasm_bytes) => Sha256::digest(wasm_bytes).to_vec(),
+                                WasmModule::ChunkedModule(ref chunked) => chunked.wasm_module_hash.clone(),
+                            };
+                            if dep.module_hash != expected_hash {
+                                return Err(UpgradeError {
+                                    stage: UpgradeStage::Installing,
+                                    reason: UpgradeErrorReason::ConcurrentChangeDetected,
+                                });
+                            }
+                        }
+                    }
```

Comment:
this check can be hardened by inspecting the origin of the change and comparing it with the expected caller (this canister running the library)

Reply:
Added in a [new PR](https://github.com/oggy-dfin/ic_call_utils/pull/2)